File: challs/idor.dir/idor.pl
#! /usr/bin/env perl

use strict;
use warnings;
use DBI;
use Mojolicious;
use Mojo::Log;
use Mojo::URL;
use Mojo::UserAgent;
use String::Random qw(random_regex);

my $driver = "SQLite";
my $dsn = "DBI:$driver:dbname=/tmp/idor.db";
my $dbh = DBI->connect($dsn, undef, undef, { AutoCommit => 1 }) or exit;

my $username = "";
my $app = Mojolicious->new;
$app = $app->log(Mojo::Log->new(path => '/dev/null'));


sub generateToken {
    #
    #    Generate a pseudo-random token
    #
    return random_regex("[0-9a-zA-Z]{64}");
}


sub getUserAccountDetails {
    #
    #    Get account details for user
    #
    my $accountId = shift;
    my $sth = $dbh->prepare("SELECT id, balance, description FROM accounts WHERE id=?");
    $sth->bind_param(1, $accountId);
    $sth->execute;
    $sth->bind_columns(\my($id, $balance, $desc));
    $sth->fetch;

    if (!$id) {
        return;
    }

    return "Account ID: $id,\nBalance:  $balance,\nDescription:  $desc";

}

######################   API   ##############################

$app->hook(before_dispatch => sub {
    #
    #    Get token from HTTP header
    #
    my $c = shift;
    my $endpoint = $c->req->url->path;
   
    if ($endpoint eq '/accounts/new'){
        return;
    }
    
    my $token = $c->req->headers->header('X-API-TOKEN');
    my $sth = $dbh->prepare("SELECT username FROM accounts WHERE token=? and username=?");
    $sth->bind_param(1, $token);
    $sth->bind_param(2, $username);
    $sth->execute;
    my $user = $sth->fetchrow_array;
    
    if (!$user) {
        return $c->render(
            status => 401,
            template => 'unauthorized',
            content => 'Invalid X-Api-Token'
         );
    }
});

$app->routes->get('/accounts/new' => sub {
    #
    #    Get account details for user
    #
    my $c = shift;
    my $token = generateToken();
    my $defaultBalance = 100;
    
    my $sth = $dbh->prepare("INSERT OR REPLACE INTO accounts(username, token, balance) VALUES(?,?,?)");
    $sth->bind_param(1, $username);
    $sth->bind_param(2, $token);
    $sth->bind_param(3, $defaultBalance);
    $sth->execute;

    $sth = $dbh->prepare("SELECT id from accounts WHERE username=?");
    $sth->bind_param(1, $username);
    $sth->execute;
    my $accountId = $sth->fetchrow;

    my $response = "Your account $accountId has been successfully created.
    Your associated token is $token
    You can view account details here /accounts/$accountId/details
    ";

    return $c->render(
        status => 200,
        text => $response
    );
});

$app->routes->get('/accounts/(:id)/details' => sub {
    #
    #    Return secret page
    #
    # Username and token already verified in function verify_token
    #
    my $c = shift;
    my $accountId = $c->param('id');
    my $accountDetails = getUserAccountDetails($accountId);
        
    if (!$accountDetails) {
        return $c->render(
            status => 404,
            template => 'not_found_cust',
            content => "Account not found"
        );
    }
    
    return $c->render(
        status => 200,
        text => $accountDetails
    );
});

####################     Main   #############################

sub main {

    $username = $ARGV[0];
    my $token = $ARGV[1];
    my $endpoint = $ARGV[2];

    if (!$username) {
        print "Missing username";
        exit 0;
    }
    
    if (!$endpoint) {
        $endpoint = '/accounts/new';
    }
    
    my $url = Mojo::URL->new($endpoint);
    my $ua = Mojo::UserAgent->new();
    $ua->server->app($app);
    my $response = $ua->get($url => {'X-API-TOKEN' => $token});
    print $response->res->content->asset->{'content'};
    $dbh->disconnect();
}

main();
1;

__DATA__

@@ forbidden.html.ep
% layout 'default';
<h1>Forbidden</h1>
<p><%= $content %></p>

@@ unauthorized.html.ep
% layout 'default';
<h1>Unauthorized</h1>
<p><%= $content %></p>

@@ not_found_cust.html.ep
% layout 'default';
<h1>Not found</h1>
<p><%= $content %></p>

@@ not_found.html.ep
% layout 'default';
<h1>Not found</h1>
<p>The requested URL was not found on the server.  If you entered the URL manually please check your spelling and try again.</p>

@@ layouts/default.html.ep
<!DOCTYPE html>
<html>
  <head><title>MyApp</title></head>
  <body><%= content %></body>
</html>

File: challs/idor.dir/idor.py
#!/usr/bin/python3

import random
import sqlite3
import string
import sys

from flask import Flask, g, request
from werkzeug.exceptions import (BadRequest,
                                 NotFound, Unauthorized)


def generate_token():
    """
        Generate a pseudo-random token
    """
    return ''.join(random.choice(string.ascii_letters + string.digits) for _ in range(64))


def get_user_account_details(account_id):
    """
        Get account details for user
    """
    g.cursor.execute("SELECT id, balance, description FROM accounts WHERE id=?", (account_id,))
    account = g.cursor.fetchone()
    if not account:
        return

    return "Account ID: {},\nBalance: {},\nDescription: {}".format(
        account[0],
        account[1],
        account[2],
    )


def create_app():
    """
        Initialize Flask application
    """
    app = Flask(__name__)

    @app.before_request
    def verify_token():
        """
            Get token from HTTP header
        """
        if request.path == '/accounts/new':
            return

        try:
            token = request.environ['HTTP_X_API_TOKEN']
        except (KeyError, IndexError, TypeError):
            raise BadRequest('Missing or invalid token')

        g.cursor.execute(
            "SELECT username FROM accounts WHERE token=? and username=?",
            (token, g.username)
        )
        user = g.cursor.fetchone()
        if not user:
            raise Unauthorized("Invalid X-Api-Token")

    @app.route('/accounts/new')
    def create_account():
        """
            Get or create an account for user
        """
        token = generate_token()
        default_balance = 100
        g.cursor.execute(
            "INSERT OR REPLACE INTO accounts(username, token, balance) VALUES(?,?,?)",
            (g.username, token, default_balance)
        )
        account_id = g.cursor.execute(
            "SELECT id from accounts WHERE username=?",
            (g.username,)
        )
        account_id = g.cursor.fetchone()[0]

        response = """Your account {} has been successfully created.
        Your associated token is {}
        You can view account details here /accounts/{}/details"""
        return response.format(account_id, token, account_id)

    @app.route('/accounts/<account_id>/details')
    def get_account_details(account_id):
        """
            Return user account details
        """
        # Username and token already verified in function verify_token

        account_details = get_user_account_details(
            account_id=account_id,
        )
        if not account_details:
            raise NotFound('Account not found')

        return account_details

    return app

APP = create_app()
APP.config['DEBUG'] = True
APP.config['TESTING'] = True


if __name__ == '__main__':

    # Parse params
    if len(sys.argv) != 4:
        print("Missing parameters")
        sys.exit(0)

    username = sys.argv[1]
    token = sys.argv[2]
    endpoint = sys.argv[3]

    if not username:
        print("Missing username")
        sys.exit(0)

    # Init app and db cursor
    try:
        tester = APP.test_client()
        ctx = APP.test_request_context()
        ctx.push()
        conn = sqlite3.connect('/tmp/idor.db', isolation_level=None)
        cursor = conn.cursor()
        g.cursor = cursor
        g.username = username
    except:
        print('Error while connecting to db.')
        sys.exit(1)

    if not endpoint:
        endpoint = '/accounts/new'

    # Make request
    response = tester.get(
        endpoint,
        headers={'X-API-TOKEN': token},
    )

    conn.commit()
    conn.close()
    print(response.get_data().decode('unicode_escape'))
